File: ddt.py
#!/usr/bin/env python3
"""DDT (radiometric snapshot) save/load for the Magnity Linux viewer.

A DDT is the camera world's "pure temperature stream" frozen to a file: it stores the
per-pixel measurement frame (radiometric counts) plus everything needed to turn any pixel
into °C later — so you can re-measure or (re)calibrate temperature offline, without the
camera. (The official Magnity .ddt is an 884-byte header + raw frame; we keep the same
"fixed header + frame" shape but use our own documented header, so these files are
self-contained for this app rather than byte-compatible with the Android SDK.)

Layout (little-endian):
    0   4   magic  b"MDDT"
    4   2   version (=1)
    6   2   W
    8   2   H
   10   2   dtype  (0 = float32 counts)
   12   4   fpa_raw  (live sensor/FPA temperature counts; 0xFFFFFFFF if unknown)
   16   8   timestamp (float64, unix seconds)
   24   1   calibrated (0/1) — whether a,b,lut below are valid
   25   1   lut_idx
   26   6   (reserved)
   32   8   cal a (float64)
   40   8   cal b (float64)
   48  208  (reserved, zero)
  256   W*H*4  frame, float32 measurement counts (row-major HxW)
"""
import struct, time, os
import numpy as np

MAGIC = b"MDDT"
VERSION = 1
HEADER = 256


def save(path, frame, fpa_raw=None, a=1.0, b=0.0, lut_idx=0, calibrated=False, ts=None):
    """Write a DDT. `frame` is HxW measurement counts (the viewer's last_frame)."""
    f = np.ascontiguousarray(np.asarray(frame, np.float32))
    H, W = f.shape
    ts = time.time() if ts is None else float(ts)
    fpa = 0xFFFFFFFF if fpa_raw is None else int(fpa_raw) & 0xFFFFFFFF
    hdr = bytearray(HEADER)
    struct.pack_into("<4sHHHH", hdr, 0, MAGIC, VERSION, W, H, 0)
    struct.pack_into("<I", hdr, 12, fpa)
    struct.pack_into("<d", hdr, 16, ts)
    struct.pack_into("<BB", hdr, 24, 1 if calibrated else 0, int(lut_idx) & 0xFF)
    struct.pack_into("<dd", hdr, 32, float(a), float(b))
    with open(path, "wb") as fh:
        fh.write(hdr)
        fh.write(f.tobytes())


def load(path):
    """Read a DDT -> dict(frame, W, H, fpa_raw, ts, calibrated, a, b, lut_idx)."""
    with open(path, "rb") as fh:
        hdr = fh.read(HEADER)
        if len(hdr) < HEADER or hdr[:4] != MAGIC:
            raise ValueError("not a DDT file (bad magic)")
        _, ver, W, H, dtype = struct.unpack_from("<4sHHHH", hdr, 0)
        fpa, = struct.unpack_from("<I", hdr, 12)
        ts, = struct.unpack_from("<d", hdr, 16)
        calibrated, lut_idx = struct.unpack_from("<BB", hdr, 24)
        a, b = struct.unpack_from("<dd", hdr, 32)
        data = fh.read(W * H * 4)
        frame = np.frombuffer(data, np.float32).reshape(H, W).copy()
    return dict(frame=frame, W=W, H=H,
                fpa_raw=(None if fpa == 0xFFFFFFFF else fpa),
                ts=ts, calibrated=bool(calibrated), a=a, b=b, lut_idx=lut_idx,
                name=os.path.basename(path))


if __name__ == "__main__":
    # round-trip self-test
    fr = (np.random.rand(120, 160).astype(np.float32) * 1000 + 28000)
    p = "/tmp/_ddt_test.ddt"
    save(p, fr, fpa_raw=19633, a=8.6, b=-28730.0, lut_idx=2, calibrated=True)
    d = load(p)
    assert d["W"] == 160 and d["H"] == 120
    assert d["fpa_raw"] == 19633 and d["lut_idx"] == 2 and d["calibrated"]
    assert np.allclose(d["frame"], fr)
    print("DDT round-trip OK:", {k: v for k, v in d.items() if k != "frame"})

File: android/app/src/main/java/com/magnity/thermalcam/data/Ddt.kt
package com.magnity.thermalcam.data

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * DDT radiometric snapshot save/load — Kotlin port of `ddt.py` (same file format, so
 * snapshots interchange with the Linux viewer).
 *
 * Layout (little-endian): see ddt.py — 256-byte header ("MDDT", version, W, H, dtype,
 * fpa_raw, timestamp f64, calibrated, lut_idx, cal a/b f64) + W*H float32 frame.
 */
object Ddt {
    private const val HEADER = 256
    private val MAGIC = byteArrayOf('M'.code.toByte(), 'D'.code.toByte(), 'D'.code.toByte(), 'T'.code.toByte())
    private const val VERSION = 1

    data class Snapshot(
        val frame: FloatArray,
        val width: Int,
        val height: Int,
        val fpaRaw: Int?,
        val timestamp: Double,
        val calibrated: Boolean,
        val a: Double,
        val b: Double,
        val lutIdx: Int,
        val name: String = "",
    )

    fun save(
        out: OutputStream, frame: FloatArray, width: Int, height: Int,
        fpaRaw: Int?, a: Double, b: Double, lutIdx: Int, calibrated: Boolean,
        timestamp: Double = System.currentTimeMillis() / 1000.0,
    ) {
        require(frame.size == width * height)
        val hdr = ByteBuffer.allocate(HEADER).order(ByteOrder.LITTLE_ENDIAN)
        hdr.put(MAGIC)
        hdr.putShort(VERSION.toShort())
        hdr.putShort(width.toShort())
        hdr.putShort(height.toShort())
        hdr.putShort(0)                                     // dtype 0 = float32 counts
        hdr.putInt(fpaRaw ?: -1)                            // 0xFFFFFFFF if unknown
        hdr.putDouble(timestamp)
        hdr.put(if (calibrated) 1 else 0)
        hdr.put((lutIdx and 0xFF).toByte())
        hdr.position(32)
        hdr.putDouble(a)
        hdr.putDouble(b)
        out.write(hdr.array())
        val px = ByteBuffer.allocate(frame.size * 4).order(ByteOrder.LITTLE_ENDIAN)
        px.asFloatBuffer().put(frame)
        out.write(px.array())
        out.flush()
    }

    fun load(input: InputStream, name: String = ""): Snapshot {
        val hdrBytes = ByteArray(HEADER)
        var off = 0
        while (off < HEADER) {
            val r = input.read(hdrBytes, off, HEADER - off)
            if (r < 0) throw IllegalArgumentException("truncated DDT header")
            off += r
        }
        if (!hdrBytes.copyOf(4).contentEquals(MAGIC)) throw IllegalArgumentException("not a DDT file (bad magic)")
        val hdr = ByteBuffer.wrap(hdrBytes).order(ByteOrder.LITTLE_ENDIAN)
        hdr.position(4)
        hdr.short // version
        val w = hdr.short.toInt() and 0xFFFF
        val h = hdr.short.toInt() and 0xFFFF
        hdr.short // dtype
        val fpa = hdr.int
        val ts = hdr.double
        val calibrated = hdr.get().toInt() != 0
        val lutIdx = hdr.get().toInt() and 0xFF
        hdr.position(32)
        val a = hdr.double
        val b = hdr.double

        val data = ByteArray(w * h * 4)
        off = 0
        while (off < data.size) {
            val r = input.read(data, off, data.size - off)
            if (r < 0) throw IllegalArgumentException("truncated DDT frame")
            off += r
        }
        val frame = FloatArray(w * h)
        ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN).asFloatBuffer().get(frame)
        return Snapshot(frame, w, h, if (fpa == -1) null else fpa, ts, calibrated, a, b, lutIdx, name)
    }
}
